src/algebra/module/bimodule.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import ring_theory.tensor_product
+
+/-!
+# Bimodules
+
+One frequently encounters situations in which several sets of scalars act on a single space, subject
+to compatibility condition(s). A distinguished instance of this is the theory of bimodules: one has
+two rings `R`, `S` acting on an additive group `M`, with `R` acting covariantly ("on the left")
+and `S` acting contravariantly ("on the right"). The compatibility condition is just:
+`(r • m) • s = r • (m • s)` for all `r : R`, `s : S`, `m : M`.
+
+This situation can be set up in Mathlib as:
+```lean
+variables (R S M : Type*) [ring R] [ring S]
+variables [add_comm_group M] [module R M] [module Sᵐᵒᵖ M] [smul_comm_class R Sᵐᵒᵖ M]
+```
+The key fact is:
+```lean
+example : module (R ⊗[ℕ] Sᵐᵒᵖ) M := tensor_product.algebra.module
+```
+Note that the corresponding result holds for the canonically isomorphic ring `R ⊗[ℤ] Sᵐᵒᵖ` but it is
+preferable to use the `R ⊗[ℕ] Sᵐᵒᵖ` instance since it works without additive inverses.
+
+Bimodules are thus just a special case of `module`s and most of their properties follow from the
+theory of `module`s`. In particular a two-sided submodule of a bimodule is simply a term of type
+`submodule (R ⊗[ℕ] Sᵐᵒᵖ) M`.
+
+This file is a place to collect results which are specific to bimodules.
+
+## Main definitions
+
+ * `subbimodule.mk`
+ * `subbimodule.smul_mem`
+ * `subbimodule.smul_mem'`
+ * `subbimodule.to_submodule`
+ * `subbimodule.to_submodule'`
+
+## Implementation details
+
+For many definitions and lemmas it is preferable to set things up without opposites, i.e., as:
+`[module S M] [smul_comm_class R S M]` rather than `[module Sᵐᵒᵖ M] [smul_comm_class R Sᵐᵒᵖ M]`.
+The corresponding results for opposites then follow automatically and since `(Sᵐᵒᵖ)ᵐᵒᵖ` is not defeq
+to `S`, this approach is more general.
```

Comment:
Are there mathematical situations where it's helpful to talk about a bimodule with both actions acting from the left?

Reply:
I guess restriction / extension of scalars is one though in that case more is true (and we'll have `is_scalar_tower` rather than `smul_comm_class`).

I guess another case should be representation theory of commutative additive groups.